java/arcs/core/data/proto/AccessPathProtoDecoder.kt

```diff
@@ -22,13 +22,41 @@ fun AccessPathProto.decode(
     requireNotNull(connectionSpec) {
         "Connection '$handleConnection' not found in connection specs!"
     }
-    val selectors = selectorsList.map {
-        when (it.selectorCase) {
-            AccessPathProto.Selector.SelectorCase.FIELD -> AccessPath.Selector.Field(it.field)
-            else -> throw IllegalArgumentException(
-                "Cannot decode a ${it.selectorCase.name} type to a [AccessPath.Selector]."
-            )
+    val selectors = selectorsList.map { it.decode() }
+    return AccessPath(particleSpec, connectionSpec, selectors)
+}
+
+fun AccessPath.encode(): AccessPathProto {
+    val proto = AccessPathProto.newBuilder()
+        .addAllSelectors(selectors.map { it.encode() })
+    when (root) {
+        is AccessPath.Root.Handle -> {
+            val handleRoot = root as AccessPath.Root.Handle
+            proto.handleConnection = handleRoot.handle.name
         }
+        is AccessPath.Root.HandleConnection -> {
+            val connectionRoot = root as AccessPath.Root.HandleConnection
```

Comment:
This is unfortunate as smart casts are not supported across packages.

Reply:
Yeah, it's annoying.. I never really understood why that doesn't work.